File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int l = 0, r = 0, globalLen = 0; 
        std::unordered_map<int, int> m;  
        
        while(r < s.length()){
            m[s.at(r)] += 1;
            while(m[s.at(r)] > 1){
                if(m[s.at(l)] > 1){
                    m[s.at(l)] -= 1;
                } else{
                    m.erase(s.at(l));
                }
                l++;
            }
            r++;
            globalLen = std::max(globalLen, r-l);
        }
        
        return globalLen;
    }
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        # make two pointers and storage
        l, r, maxLen = 0, 0, 0
        storage = collections.Counter()
        
        # increment right pointer and add char to storage
        while r < len(s):
            
            storage[s[r]] += 1
            
            # increment left pointer until no more duplicate values
            while storage[s[r]] > 1: 
                if storage[s[l]] > 1: 
                    storage[s[l]] -= 1
                else: 
                    del storage[s[l]]
                l += 1
                
            r += 1
            
            # keep track of maximum length substring without repeating characters
            maxLen = max(maxLen, r-l)
                
        return maxLen
            
        